File: sap-rpt-1/code_samples/python/predict_parquet.py
import json
from pathlib import Path

import requests

# API Configuration
# Note: Replace the placeholder "XXXXXXXXXX" values with your actual configuration.
API_URL = "XXXXXXXXXX/predict_parquet"  # can refer to the readme for the exact format
AI_RESOURCE_GROUP = "XXXXXXXXXX"  # e.g., "default"

# OAuth Configuration
# You can get the credentials from your BTP service key.
# Consider using environment variables and ensure handling secrets in a secure way.
OAUTH_CONFIG = {
    "auth_url": "XXXXXXXXXX/oauth/token",
    "client_id": "XXXXXXXXXX",
    "client_secret": "XXXXXXXXXX",
}


def get_access_token():
    """Get OAuth2 access token using client credentials flow."""
    token_data = {
        "grant_type": "client_credentials",
        "client_id": OAUTH_CONFIG["client_id"],
        "client_secret": OAUTH_CONFIG["client_secret"],
    }

    response = requests.post(
        OAUTH_CONFIG["auth_url"],
        data=token_data,
        headers={"Content-Type": "application/x-www-form-urlencoded"},
    )

    if response.status_code == 200:
        return response.json()["access_token"]
    else:
        raise Exception(
            f"Failed to get access token: {response.status_code} - {response.text}"
        )


def predict_parquet_file(access_token, file_path):
    """
    Upload a parquet file and make predictions for multiple columns.

    Args:
        access_token: OAuth2 access token
        file_path: Path to the parquet file

    Returns:
        API response as dictionary
    """
    # Configure predictions for 'stock' columns with PLACEHOLDER
    prediction_config = {
        "target_columns": [
            {"name": "category", "prediction_placeholder": "PLACEHOLDER"}
        ]
    }

    headers = {
        "Authorization": f"Bearer {access_token}",
        "AI-Resource-Group": AI_RESOURCE_GROUP,
    }

    with open(file_path, "rb") as f:
        files = {"file": (Path(file_path).name, f, "application/octet-stream")}

        data = {
            "prediction_config": json.dumps(prediction_config),
            "parse_data_types": "false",
        }

        response = requests.post(API_URL, headers=headers, files=files, data=data)

    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(f"Request failed: {response.status_code} - {response.text}")


if __name__ == "__main__":
    HERE = Path(__file__).parent
    parquet_file = HERE.parent / "data" / "product_data.parquet"

    try:
        # Step 1: Get access token
        print("Obtaining access token...")
        access_token = get_access_token()
        print("Access token obtained successfully!")

        # Step 2: Make prediction with parquet file
        print(f"Uploading {parquet_file} and making predictions...")
        result = predict_parquet_file(access_token=access_token, file_path=parquet_file)

        print("Prediction successful!")
        print(json.dumps(result, indent=2))

    except FileNotFoundError:
        print(f"Error: File '{parquet_file}' not found")
    except Exception as e:
        print(f"Error: {e}")

File: sap-rpt-1/code_samples/java/PredictParquet.java
import org.apache.hc.client5.http.classic.methods.HttpPost;
import org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder;
import org.apache.hc.client5.http.entity.UrlEncodedFormEntity;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.hc.core5.http.ContentType;
import org.apache.hc.core5.http.NameValuePair;
import org.apache.hc.core5.http.io.entity.EntityUtils;
import org.apache.hc.core5.http.message.BasicNameValuePair;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

class PredictParquet {

    static final String API_URL = "XXXXXXXXXX/predict_parquet";  // can refer to the readme for the exact format
    static final String AI_RESOURCE_GROUP = "XXXXXXXXXX";  // e.g., "default"

    // OAuth Configuration
    // You can get the credentials from your BTP service key.
    // Consider using environment variables and ensure handling secrets in a secure way.
    static final String AUTH_URL = "XXXXXXXXXX/oauth/token";
    static final String CLIENT_ID = "XXXXXXXXXX";
    static final String CLIENT_SECRET = "XXXXXXXXXX";

    public static String getAccessToken() throws Exception {
        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
            HttpPost httpPost = new HttpPost(AUTH_URL);

            List<NameValuePair> params = new ArrayList<>();
            params.add(new BasicNameValuePair("grant_type", "client_credentials"));
            params.add(new BasicNameValuePair("client_id", CLIENT_ID));
            params.add(new BasicNameValuePair("client_secret", CLIENT_SECRET));
            httpPost.setEntity(new UrlEncodedFormEntity(params, StandardCharsets.UTF_8));

            try (CloseableHttpResponse response = httpClient.execute(httpPost)) {
                String responseBody = EntityUtils.toString(response.getEntity());

                if (response.getCode() == 200) {
                    Gson gson = new Gson();
                    JsonObject jsonObject = gson.fromJson(responseBody, JsonObject.class);
                    return jsonObject.get("access_token").getAsString();
                } else {
                    throw new Exception("Failed to get access token: " + response.getCode() + " - " + responseBody);
                }
            }
        }
    }

    public static void main(String[] args) {
        try {
            Path parquetFile = Paths.get("../data/product_data.parquet").toAbsolutePath().normalize();
            System.out.println("Obtaining access token...");
            String accessToken = getAccessToken();
            System.out.println("Access token obtained!");

            try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
                HttpPost httpPost = new HttpPost(API_URL);

                httpPost.setHeader("Authorization", "Bearer " + accessToken);
                httpPost.setHeader("AI-Resource-Group", AI_RESOURCE_GROUP);

                var entity = MultipartEntityBuilder.create()
                    .addTextBody("prediction_config",
                        "{\"target_columns\":[{\"name\":\"category\",\"prediction_placeholder\":\"PLACEHOLDER\",\"task_type\":\"classification\",\"top_k\":1}]}",
                        ContentType.APPLICATION_JSON)
                    .addTextBody("parse_data_types", "false")
                    .addBinaryBody("file", parquetFile.toFile(), ContentType.DEFAULT_BINARY, parquetFile.getFileName().toString())
                    .build();

                httpPost.setEntity(entity);

                try (CloseableHttpResponse response = httpClient.execute(httpPost)) {
                    String responseBody = EntityUtils.toString(response.getEntity());

                    if (response.getCode() == 200) {
                        System.out.println("Prediction successful!");
                        System.out.println(responseBody);
                    } else {
                        System.err.println("Request failed: " + response.getCode());
                        System.err.println(responseBody);
                    }
                }
            }
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
